src/pcp.rs

```diff
@@ -429,167 +472,219 @@ impl<F: FieldElement> From<Vec<F>> for Proof<F> {
     }
 }
 
-/// Generate the verifier for an input and proof (or the verifier share for an input share and
-/// proof share).
+/// Generate a verifier message for an input and proof (or the verifier share for an input share
+/// and proof share).
 ///
 /// Parameters:
 /// * `x` is the input.
 /// * `pf` is the proof.
 /// * `query_rand` is the verifier's randomness.
 /// * `joint_rand` is the randomness shared by the prover and verifier.
-pub fn query<F, G, V>(
+pub fn query<F, V>(
     x: &V,
     pf: &Proof<F>,
     query_rand: &[F],
     joint_rand: &[F],
 ) -> Result<Verifier<F>, PcpError>
 where
     F: FieldElement,
-    G: Gadget<F>,
-    V: Value<F, G>,
+    V: Value<F>,
 {
-    if query_rand.len() != 1 {
-        return Err(PcpError::Query("incorrect length of verifier randomness"));
-    }
-
     let g_calls = x.valid_gadget_calls();
-    let m = (g_calls + 1).next_power_of_two();
-    let g = x.gadget(m);
-    let l = g.call_in_len();
 
-    // Allocate space for the verifier data. This includes the output of the validity circuit, the
-    // intermediate polynomials evaluated at the query randomness `r`, and the proof polynomial
-    // evaluated at `r`.
-    let mut verifier_data = Vec::with_capacity(2 + l);
+    let mut proof_len = 0;
+    let mut shim = x
+        .gadget()
+        .into_iter()
+        .enumerate()
+        .map(|(idx, g)| {
+            if idx >= query_rand.len() {
+                return Err(PcpError::Query("short query randomness"));
+            }
+
+            let g_deg = g.deg();
+            let g_arity = g.arity();
+            let m = (1 + g_calls[idx]).next_power_of_two();
 
-    // Run the validity circuit with a "shim" gadget that records each input to each gadget.
-    // Record the output of the circuit.
+            // Compute the length of the sub-proof corresponding to the `idx`-th gadget.
+            let next_len = g_arity + g_deg * (m - 1) + 1;
+            if proof_len + next_len > pf.data.len() {
+                return Err(PcpError::Query("short proof"));
+            }
+
+            let proof_data = &pf.data[proof_len..proof_len + next_len];
+            proof_len += next_len;
+
+            QueryShimGadget::new(g, query_rand[idx], proof_data, g_calls[idx])
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    if proof_len < pf.data.len() {
+        return Err(PcpError::Query("long proof"));
+    }
+
+    if query_rand.len() > shim.len() {
+        return Err(PcpError::Query("long joint randomness"));
+    }
+
+    // Create a buffer for the verifier data. This includes the output of the validity circuit as
+    // well as the intermediate polynomials evaluated at the query randomness `query_rand[idx]`,
+    // and the proof polynomial evaluated at `query_rand[idx]` for the `idx`-th gadget.
+    let data_len = 1
+        + (0..shim.len())
+            .map(|idx| shim[idx].arity() + 1)
+            .reduce(|a, b| a + b)
+            .unwrap();
```

Comment:
```suggestion
            .sum();
```
as in `prove()`, to avoid, `unwrap()`

Reply:
Done. (This also needed a type declaration, i.e., `.sum::<usize>()`.)